text/0012-jsonfile.md

```diff
@@ -0,0 +1,60 @@
+# A brief analysis of JSON file-backed storage
+
+Several components in Firefox — including libpref, XULStore, autofill, and logins — follow a general pattern:
+
+- Store data in memory, usually as a JS collection.
+- Load that data from disk during initialization.
+- Persist that data to disk in its entirety, usually serialized as JSON, at some point after each write, and/or at shutdown.
+
+[JSONFile.jsm](https://dxr.mozilla.org/mozilla-central/source/toolkit/modules/JSONFile.jsm) is a centralized implementation of this pattern. It supports Promise-based asynchronous loading, synchronous loading, and flushes asynchronously, defaulting to 1500ms after each change. libpref flushes after 500msec.
+
+This approach is well suited to data that:
+
+- Changes infrequently.
+- Is relatively small.
+- Is only ever directly read or written by a single thread.
+- Needs to be kept entirely in memory for reads.
+- Has a straightforward JSON encoding (consider the expense of base64 + UTF-8 for images).
+- Has limited durability requirements.
+
+Advantages of this approach:
+- It's relatively simple, with serialization and deserialization being the only steps.
+- Because the file is rewritten frequently, there is some hard-to-measure resilience to gradual file corruption.
+- The file on disk is human-readable and editable, which is convenient for development, test, and support.
+- The in-memory data can be queried and manipulated synchronously.
+- Development is 'natural' for front-end developers — get things working with only in-memory data, then add persistence on top.
+
+Disadvantages:
+- Versioning is often an afterthought: engineers rarely think to version in-memory data formats.
+- Change tracking is almost always an afterthought, thanks to the 'natural' development process: adding syncing later becomes difficult.
+- Users can edit the file on disk, which can result in persistent state that was never created by the component's own logic.
+- Frequent writes will cause the entire file to be written to disk repeatedly, [which causes complaints about SSD impact](https://bugzilla.mozilla.org/show_bug.cgi?id=1304389).
+- The entire file typically needs to be read into memory to be used, which increases memory footprint and can impact startup time.
+- Writes at shutdown harm the user experience and don't happen during a crash.
+- During persistence, the entire file exists twice on disk, which could be problematic on mobile platforms for large data sets.
+- The only approach to cross-process use is duplication of all or part of the in-memory data, which can increase memory footprint. It's not possible for multiple processes to collaborate on the same data via the filesystem. This is why Firefox for Android still uses the SQLite implementation of `nsILoginManagerStorage`.
```

Comment:
Again, explain your conclusion: "... still uses the SQLite implementation of ..., when it would be better (reasons?) to use an Android-native storage solution."